File: solutions/python/triangle/1/triangle.py
def validate(sides):
    if (sides[0] + sides[1] + sides[2]) > 0:
        return (sides[0] + sides[1] >= sides[2]) and (sides[1] + sides[2] >= sides[0]) and (sides[0] + sides[2] >= sides[1])
    return False


def equilateral(sides):
    if validate(sides):
        return sides[0] == sides[1] == sides[2]
    return False


def isosceles(sides):
    if validate(sides):
        return sides[0] == sides[1] or sides[1] == sides[2] or sides[2] == sides[0]
    return False
    

def scalene(sides):
    if equilateral(sides) or isosceles(sides):
        return False
    return validate(sides)

File: solutions/cpp/triangle/3/triangle.cpp
#include "triangle.h"
#include <stdexcept>

namespace triangle {

bool positive_sides(double a, double b, double c) {
  return (a > 0) && (b > 0) && (c > 0);
}

bool valid_triangle(double a, double b, double c) {
  return (a + b >= c) && (b + c >= a) && (c + a >= b);
}

flavor kind(double a, double b, double c) {
  if (!positive_sides(a, b, c) || !valid_triangle(a, b, c))
    throw std::domain_error("not a triangle");

  if ((a == b) && (b == c))
    return triangle::flavor::equilateral;
  else if ((a == b) || (b == c) || (a == c))
    return triangle::flavor::isosceles;
  else
    return triangle::flavor::scalene;
};

} // namespace triangle
